warp10/src/main/java/io/warp10/script/functions/DTW.java

```diff
@@ -205,77 +391,159 @@ public double compute(double[] values1, int offset1, int len1, double[] values2,
+
+          bestPreviousTotalDistance = Math.min(left, Math.min(bottom, bottomLeft));
+        }
+
+        // Avoid computing the distance if the best total previous distance is already over the threshold.
+        if (threshold >= bestPreviousTotalDistance) {
+          //
+          // Compute distance.
+          // DTW simply considers the delta in values, not the delta in indices
+          //
+
+          double d = distance.measure(values1, j, values2, i);
+
+          b[j] = d + bestPreviousTotalDistance;
+
+          if (!belowThreshold && b[j] <= threshold) {
+            belowThreshold = true;
+          }
+        } else {
+          b[j] = Double.POSITIVE_INFINITY;
         }
       }
-      
+
       // Exit if no value is below threshold
       if (!belowThreshold) {
-        break;
+        return -1.0D;
       }
-      
+
       // Shift b into a
       double[] tmp = a;
-      a = b;           
+      a = b;
       b = tmp;
     }
-    
-    if (!belowThreshold) {
-      return -1.0D;
+
+    // Result is the top right value of the match matrix.
+    // It may be over the threshold if a[k] <= threshold for some k<(len1-1);
+    if (a[len1 - 1] <= threshold) {
+      return a[len1 - 1];
+    } else {
+      return -1;
+    }
+  }
+
+  // Functional interface defining a distance function used by DTW.
+  @FunctionalInterface
+  public interface DTWDistance {
+    double measure(double[][] values1, int index1, double[][] values2, int index2) throws WarpScriptException;
+  }
+
+  public static double manhattan(double[][] values1, int index1, double[][] values2, int index2) throws WarpScriptException {
+    double d = 0D;
+    for (int dimension = 0; dimension < values1.length; dimension++) {
+      d += Math.abs(values1[dimension][index1] - values2[dimension][index2]);
     }
-    
-    double dtwDist = Double.POSITIVE_INFINITY;
-    
-    for (int i = 0; i < a.length; i++) {
-      if (a[i] < dtwDist) {
-        dtwDist = a[i];
-      }
+    return d;
+  }
+
+  public static double euclidean(double[][] values1, int index1, double[][] values2, int index2) throws WarpScriptException {
+    double d = 0D;
+    for (int dimension = 0; dimension < values1.length; dimension++) {
+      d += Math.pow(values1[dimension][index1] - values2[dimension][index2], 2.0D);
     }
-    
-    if (!manhattan) {
-      dtwDist = Math.sqrt(dtwDist);
+    return Math.sqrt(d);
+  }
+
+  public static double loxodromic(double[][] values1, int index1, double[][] values2, int index2) throws WarpScriptException {
+    if (2 != values1.length) {
+      throw new WarpScriptException("Loxodromic distance must be given two dimensions: lat, lon.");
+    }
+
+    double lat1 = values1[0][index1];
+    double lat2 = values2[0][index2];
+    double lon1 = values1[1][index1];
+    double lon2 = values2[1][index2];
+
+    return GeoXPLib.loxodromicDistance(GeoXPLib.toGeoXPPoint(lat1, lon1), GeoXPLib.toGeoXPPoint(lat2, lon2));
```

Comment:
You should re-implement the distance computation directly on lat/lon instead of converting to HHCode then letting loxodromicDistance split the HHCode. Same goes for the orthodromic distance below.